### ISSUE TITLE
Error compiling Semigroups package with libsemigroups 0.4.0

### ISSUE THREAD

**Issue by wilfwilson** [gap-packages/Semigroups#380]
Everything is working fine on Travis, which is great. Unfortunately, I get the following errors when compiling the Semigroups package with the new libsemigroups version 0.4.0:
```
  CXX      src/semigroups_la-semigrp.lo
In file included from src/semigrp.cc:19:
In file included from ./src/semigrp.h:25:
./src/libsemigroups/src/semigroups.h:1269:43: error: default initialization of
      an object of const type 'const libsemigroups::Semigroup::IteratorMethods'
      requires a user-provided default constructor
  C const Semigroup::iterator_base<T, C>::_methods;
                                          ^
./src/libsemigroups/src/semigroups.h:988:16: note: in instantiation of static
      data member 'libsemigroups::Semigroup::iterator_base<const
      libsemigroups::Element *,
      libsemigroups::Semigroup::IteratorMethods>::_methods' requested here
        return _methods.indirection(_it_vec);
               ^
src/semigrp.cc:100:50: note: in instantiation of member function
      'libsemigroups::Semigroup::iterator_base<const libsemigroups::Element *,
      libsemigroups::Semigroup::IteratorMethods>::operator*' requested here
    SET_ELM_PLIST(out, i++, converter->unconvert(*it));
                                                 ^
src/semigrp.cc:542:12: note: in instantiation of function template
      specialization
      'iterator_to_plist<libsemigroups::Semigroup::iterator_base<const
      libsemigroups::Element *, libsemigroups::Semigroup::IteratorMethods> >'
      requested here
    return iterator_to_plist(
           ^
In file included from src/semigrp.cc:19:
In file included from ./src/semigrp.h:25:
./src/libsemigroups/src/semigroups.h:1269:43: error: default initialization of
      an object of const type 'const
      libsemigroups::Semigroup::IteratorMethodsPairFirst' requires a
      user-provided default constructor
  C const Semigroup::iterator_base<T, C>::_methods;
                                          ^
./src/libsemigroups/src/semigroups.h:988:16: note: in instantiation of static
      data member 'libsemigroups::Semigroup::iterator_base<std::__1::pair<const
      libsemigroups::Element *, unsigned long>,
      libsemigroups::Semigroup::IteratorMethodsPairFirst>::_methods' requested
      here
        return _methods.indirection(_it_vec);
               ^
src/semigrp.cc:564:52: note: in instantiation of member function
      'libsemigroups::Semigroup::iterator_base<std::__1::pair<const
      libsemigroups::Element *, unsigned long>,
      libsemigroups::Semigroup::IteratorMethodsPairFirst>::operator*' requested
      here
      SET_ELM_PLIST(out, i++, converter->unconvert(*it));
                                                   ^
2 errors generated.
make: *** [src/semigroups_la-semigrp.lo] Error 1
```
[It seems like people have come across similar problems elsewhere](https://www.google.co.uk/search?q=error%3A+default+initialization+of+an+object+of+const+type+requires+a+user-provided+default+constructor). Since I don't understand C++, I don't really know what's going on, but I know that I can get rid of the errors and compile the Semigroups package by deleting lines `1268-9` of `libsemigroups/src/semigroups.h`:
```
  template <typename T, typename C>
  C const Semigroup::iterator_base<T, C>::_methods;
```
I'm compiling with:
```
Apple LLVM version 6.0 (clang-600.0.57) (based on LLVM 3.5svn)
Target: x86_64-apple-darwin13.4.0
Thread model: posix
```

**Comment by james-d-mitchell** [comment]
I guess when you start GAP after deleting those lines that you then get a linking failure when you start GAP, right?

**Comment by wilfwilson** [comment]
No, it loads fine and `SemigroupsTestStandard();` runs normally.

**Comment by james-d-mitchell** [comment]
Hmm, that does not make sense. The real problem is that you are using an old version of clang, but I think this should be easy enough to resolve. It is also an issue with libsemigroups rather than the Semigroups package for GAP. 

What happens if you add the line:

          IteratorMethods() {}

immediately after line 1000 in `libsemigroups/src/semigroups.h` (inside the declaration of the `struct IteratorMethods`) and also in line 1012 (inside the declaration of `struct IteratorMethodsFirstPair`):

      IteratorMethodsPairFirst() {}

??

**Comment by wilfwilson** [comment]
That totally fixes it.

I made the issue here since this error happened for me when compiling the Semigroups package - libsemigroups compiled fine. It's hard to know where to file an issue in these situations.

I guess the C++ language specification that the error is complaining about is no longer a requirement?

I realise I'm using a two year old compiler, and I don't expect it to work for ever. However the only trouble that I've had with it is this issue, and problems getting benchmarking to work with libsemigroups. In particular GAP compiles fine for me, as does every other GAP package that requires compilation.

For Mac users at least, updating clang may require upgrading to a new OS, which a user may not want or be able to do. I'm not saying you should support ancient compilers, but I think it makes sense to resolve minor issues like this one, where possible, in order to make the package as usable as possible.

I really appreciate your help in finding a fix for this. Thanks!

**Comment by james-d-mitchell** [comment]
No, worries, it's interesting that libsemigroups compiles, but not the Semigroups package. I will try to make another release of libsemigroups with these changes included later today. Or even better, can you make a PR with the changes?

**Comment by wilfwilson** [comment]
Sure.

**Comment by james-d-mitchell** [comment]
Thanks!

**Comment by james-d-mitchell** [comment]
This is resolved by Version 0.4.1 of libsemigroups.
